plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/editors/featuremodel/actions/CollapseAction.java

```diff
@@ -66,41 +79,150 @@ public void selectionChanged(SelectionChangedEvent event) {
 		}
 	};
 
-	public CollapseAction(Object viewer, IGraphicalFeatureModel graphicalFeatureModel) {
+	public CollapseAction(Object viewer, IGraphicalFeatureModel graphicalFeatureModel, IFeatureModel featureModel) {
 		super(COLLAPSE_FEATURE, viewer);
 		this.graphicalFeatureModel = graphicalFeatureModel;
+		
 		setEnabled(false);
 		if (viewer instanceof GraphicalViewerImpl) {
 			((GraphicalViewerImpl) viewer).addSelectionChangedListener(listener);
 		} else {
 			((TreeViewer) viewer).addSelectionChangedListener(listener);
 		}
 	}
+	
+	private boolean isEveryFeatureCollapsed() {
+		for (IGraphicalFeature tempFeature : graphicalFeatureArray) {
+			if (!(tempFeature.isCollapsed())) {
+				return false;
+			}
+		}
+		return true;
+	}
+	
+	
+	private void refreshGraphicalFeatures() {
+		ArrayList<IGraphicalFeature> tempGraphicalFeatureList = new ArrayList<>();
+		
+		for (IFeature tempFeature : featureArray) {
+			tempGraphicalFeatureList.add(graphicalFeatureModel.getGraphicalFeature(tempFeature));
+		}
+		
+		graphicalFeatureArray = tempGraphicalFeatureList.toArray(new IGraphicalFeature[tempGraphicalFeatureList.size()]);
+	}
+	
+	@Override
+	protected void selectionElementChanged(boolean validSelection) {
+		if (featureArray != null) {
+			for (Object obj : featureArray) {
+				((IFeature)obj).removeListener(this);
+			}
+		}
+		if (validSelection) {			
+			featureArray =  getSelectedFeatures();
+			refreshGraphicalFeatures();
+			for (IFeature tempFeature : featureArray) {
+				tempFeature.addListener(this);
+			}
+			updateProperties();
+		} else {
+			featureArray = null;
+			setEnabled(false);
+		}
+	}
 
 	@Override
 	public void run() {
-
-		final IGraphicalFeature graphicalFeature = graphicalFeatureModel.getGraphicalFeature(feature);
-		setChecked(graphicalFeature.isCollapsed());
-		// setChecked(feature.getStructure().isCollapsed());
-		final SetFeatureToCollapseOperation op = new SetFeatureToCollapseOperation(feature, graphicalFeatureModel);
+		setChecked(isEveryFeatureCollapsed());
+		changeCollapsedStatus(isEveryFeatureCollapsed());
+	}
+	
+	private void changeCollapsedStatus(boolean allCollapsed) {
+		final SetFeatureToCollapseOperation op = new SetFeatureToCollapseOperation(featureArray, graphicalFeatureModel, allCollapsed, getStringLabel());
 
 		try {
 			PlatformUI.getWorkbench().getOperationSupport().getOperationHistory().execute(op, null, null);
 		} catch (final ExecutionException e) {
 			FMUIPlugin.getDefault().logError(e);
-
+	
+		}
+	}
+	
+	private String getStringLabel() {
+		if (featureArray.length == 1) {
+			if (isEveryFeatureCollapsed()) {
+				return SET_FEATURE_EXPANDED;
+			} else {
+				return SET_FEATURE_COLLAPSED;
+			}
+		} else {
```

Comment:
What happens if the featureArray.length is 0?

Reply:
the action won't even show up in the menu then.